### PR TITLE
Fix momentum setting in BatchNorm forward pass.

### PR DESCRIPTION
**Pull request by spandantiwari**
This is a fix for issue https://github.com/pytorch/pytorch/issues/18525. The issue is related not only to ONNX export, but can manifest in other scenarios.
An existing test point in test/onnx/test_operators.py has been updated to cover this scenario as well.

### CODE REVIEW

**Comment by spandantiwari** [comment]
@houseroad @zrphercule - this is a fix for one of the ONNX issues https://github.com/pytorch/pytorch/issues/18525. Not labeling this ONNX in title as this is not just for ONNX export but more generic (I think).

**Review by ssnl — APPROVED** [review]
LGTM. Let's wait for CI.

File: torch/nn/modules/batchnorm.py [hunk 1]
```diff
@@ -60,7 +60,7 @@ def _check_input_dim(self, input):
     def forward(self, input):
         self._check_input_dim(input)
 
-        exponential_average_factor = 0.0
+        exponential_average_factor = 0.0 if self.momentum is None else self.momentum
```
Comment (houseroad) [inline torch/nn/modules/batchnorm.py]: `expected a value of type float for argument 'momentum' but found float?` error in CI. you probably also want to check whether self.momentum is None or not.

Comment (spandantiwari) [reply torch/nn/modules/batchnorm.py]: Fixed. Issue with the use of ternary operator. See comment below.

**Comment by spandantiwari** [comment]
@houseroad - A couple of failing tests just needed the .expect files to be updated. There were other failures in a few test_jit.py testpoints. There was a deeper issue, where the ScriptModule that is generated against the BatchNorm operator does not correctly capture the ternary operator `exponential_average_factor = 0.0 if self.momentum is None else self.momentum`. I have used a simpler if-else for this change, which gets the failing tests to pass. Shall I open a jit related issue for the ternary op?

**Review by houseroad — APPROVED** [review]
Looks good. Linter issue is unrelated.

**Review by apaszke — CHANGES_REQUESTED** [review]
This changes the semantics of the batch norm module in our frontend

**Comment by spandantiwari** [comment]
Hi @apaszke - Yes - while this issue was found through an ONNX model, this PR is intended to fix a possible bug in batchnorm.py itself (assuming that's what you mean by front end) for the inference mode.

**Comment by SsnL** [comment]
@apaszke I don't know why you say that this changes the behavior. Although I don't know why we need this thing, but it doesn't change any frontend behavior.

+ `training & track_running_stats`: nothing changed
+ `!training & track_running_stats`: nothing changed because `exponential_average_factor` isn't used in eval mode.
+ `training & !track_running_stats`: nothing changed because `exponential_average_factor` isn't used when `!track_running_stats`.
+ `!training & !track_running_stats`: nothing changed because `exponential_average_factor` isn't used when `!track_running_stats`

**Comment by spandantiwari** [comment]
The issue is seen when exporting the BN mode to ONNX, as outlined in https://github.com/pytorch/pytorch/issues/18525. Any momentum value of the BN node is _lost_ when translating to the ONNX graph. It is not exactly wrong, but folks find it unintuitive.

**Comment by spandantiwari** [comment]
The issue is seen when exporting the BN mode to ONNX, as outlined in https://github.com/pytorch/pytorch/issues/18525. Any momentum value of the BN node is _lost_ when translating to the ONNX graph, which is a bit unintuitive.

**Review by apaszke — APPROVED** [review]
Duh, those flags are weird and redundant. We should just use the factor to determine whether we should update the stats or not.

Anyway, this deserves a comment the source, because it is a place that's very likely to get refactored back to the simpler version in the future.

**Comment by spandantiwari** [comment]
OK, will add a comment.
